translated: docs/developer-docs/integrations/http_requests

hokosugi opened the pull request:
疑問点・不明点

今回もかなり手強かったです。修正点が多いときはこちらで修正します。

index.md
* １０行目 `abstracted away` をｃ「抽象化」と訳しました。

http_requests-how-it-works.md
* １４行目 `shim` はアダプター内にあるらしいですけど、[フォーラムに１箇所](https://forum.dfinity.org/t/release-notes-for-new-replica-version-8cae4/11685)しかヒットしないので、そのままにしました
* ２６行目 `This guarantees that on every replica the exact same response (or none at all) is used for canister execution` の括弧の中をどう訳するのが良いか悩みます。「これにより、すべてのレプリカにおいて、全く同じレスポンス（または何も違わない）が Canister の実行に使用されることが保証され、」としましたが、強調したいだけで完全に削除しても良いのかも知れません。
* ３３行目 `It is assumed that the HTTP service is honest, otherwise, it can provide any responses to any of the calling replicas of the subnet.` を「HTTP サービスは正直であると仮定します。そうでなければ，サブネットのどの呼び出し側レプリカに対しても，任意のレスポンスを提供することができません。」としました。文意は合ってるはずですが、otherwise で否定形のnot, noが後半にないので否定的に書いて良いものかわかりませんでした。
* ３５行目 `A dishonest server can easily make requests fail or provide wrong data. Also, 1/31/3 or more of dishonest subnet replicas can make requests fail.` の easily をどう変化するのか分かりにくい。「不誠実なサーバー（レプリカ）は、簡単にリクエストを失敗させたり、間違ったデータを提供したりすることになります。また、$1/3$ 以上の不誠実なサブネットレプリカでリクエスト自体を失敗させることができます。」としました。
* 59行目 ここは全体的によく分からず。` (Xnet) request` と ` Xnet requests` の違いがあるので、オラクル使うとレイテンシーがアウトコールより大きく、サブネット間（Xnet）だと更に大きくなるから、「IC のサブネットアーキテクチャの原則とやや矛盾」になると解釈しました。
* ７３行目移行 ` compromised replica` を「不正なレプリカ」としました。危殆化とかあると思いますが、これでどうでしょうか？
* １１６行目 `Manually,` を「手動」にしました。原文のままのほうが良いかも？と思いましたが、このレシピの意味がハッキリ分かっていないです。

`&ldquo;手動&rdquo;` の ”手動” をHTMLのコード（エンティティ）の `&#12300; 手動 &#12301;`にしたほうがいいかもしれません。これ以外に複数あります。

以上です。
宜しくおねがいします。

hokosugi committed:
translated: docs/developer-docs/integrations/http_requests

hoosan commented on docs/developer-docs/integrations/http_requests/index.md:
```diff
@@ -1,3 +1,35 @@
+
+これまでブロックチェーンは孤立した存在であり、スマートコントラクトが外部のサーバーや他のブロックチェーンと直接通信することはできませんでした。その理由は、ブロックチェーンは複製されたステートマシンであり、各ノードはラウンドごとに同じステートに対して同じ計算を行い、同じ遷移を行う必要があるからです。外部サービスの結果を入力として計算を行うことは、ナイーブに行うとノードのステート不一致につながりやすいため、動作させるためには技術的な配慮が必要です。
+
+Internet Computer の _HTTP(S) リクエストの登録_または _HTTP(S) アウトコール_ の機能により、ブロックチェーン史上初めて、スマートコントラクトがブロックチェーンの外部の HTTP(S) サーバーを直接呼び出し、レスポンスをスマートコントラクトの処理で使用し、その入力を使って安全に複製ステートを更新することができるようになりました。これまでのところ、スマートコントラクトと外部サーバーとの通信手段は、いわゆる*オラクル*を介したものだけでした。このドキュメントの残りの部分では、基礎となるプロトコルを参照する *HTTP* と *HTTPS* を *HTTP* に統一して使用することに注意してください。近年、公共ネットワーク上の実質的にすべての HTTP トラフィックは、保護された HTTPS 上で実行されます。
```
- canister と register の誤読と思われます
- may use *HTTP* は、「統一して使う」ではなく、「HTTPをHTTPSの意味で使ってる箇所があるかもしれないけど許してね」くらいの意味かと思います。事実、下の記載でHTTPSとしている箇所もそれなりにあります。
```suggestion
Internet Computer の _Canister HTTP(S) リクエスト_または _HTTP(S) アウトコール_ の機能により、ブロックチェーン史上初めて、スマートコントラクトがブロックチェーンの外部の HTTP(S) サーバーを直接呼び出し、レスポンスをスマートコントラクトの処理で使用し、その入力を使って安全に複製ステートを更新することができるようになりました。これまでのところ、スマートコントラクトと外部サーバーとの通信手段は、いわゆる*オラクル*を介したものだけでした。このドキュメントの残りの部分では、基礎となるプロトコルを参照する *HTTP* と *HTTPS* を代表して *HTTP* を用いることがあることに注意してください。近年、公共ネットワーク上の実質的にすべての HTTP トラフィックは、保護された HTTPS 上で実行されます。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+* このリクエストはサブネットで複製されたステートに一時的に保存されます。
+* 定期的に（ラウンドごとに）、各レプリカのネットワーキング層にある*アダプター*は、複製されたステートから保留中の HTTP アウトコールを取得します（実際には、レプリカプロセスの内部にあるアダプター層（&lsquo;shim&rsquo;）がこれを行います。アダプター自体は、セキュリティ上の理由から、別の OS レベルのプロセスにサンドボックス化されています）。
+* 各レプリカ上のアダプターは、HTTP リクエストをターゲット・サーバーに送信して実行します。
+* サーバーからの HTTP レスポンスは、サブネットの各レプリカ上のアダプターが受信し、レプリカプロセスのコンポーネントに提供されます。アダプターは、ネットワークレスポンスサイズを `max_response_bytes` に制限しています。この値はリクエストの価格に影響するので、期待するレスポンスに対してできる限り低く設定することが重要です。価格は `max_response_bytes` のサイズとともに増加し、実際のレスポンスサイズは考慮されず、最大値のみが考慮されます。
```
which defaults to $2$ MB and can be set to lower values の部分の訳が抜けているようです
```suggestion
* サーバーからの HTTP レスポンスは、サブネットの各レプリカ上のアダプターが受信し、レプリカプロセスのコンポーネントに提供されます。アダプターは、ネットワークレスポンスサイズを `max_response_bytes` に制限しており、デフォルトでは $2$ MB に設定されていますが、より小さい値を設定することもできます。この値はリクエストの価格に影響するので、期待するレスポンスに対してできる限り低く設定することが重要です。価格は `max_response_bytes` のサイズとともに増加し、実際のレスポンスサイズは考慮されず、最大値のみが考慮されます。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+
+要約すると、HTTP リクエストを実行するために、各レプリカは、外部の Web サーバーから受け取った HTTP レスポンスの（オプションで変換された）インスタンスを Internet Computer のコンセンサス層を介してプッシュします。サブネットのレプリカは、レプリカが受け取ったすべてのサーバーレスポンスに基づいて、Canister に提供するレスポンスに合意できるようにします。オプションの変換により、サーバーから各レプリカで受信したレスポンスが部分的に異なる場合、それらの違いは排除され、同じ変換されたレスポンスがすべての（正直な）レプリカのコンセンサスに提供されることが保証されます。これにより、すべてのレプリカにおいて、全く同じレスポンス（または何も違わない）が Canister の実行に使用されることが保証され、それによってこの機能を使用するときに不一致が起こらず、サブネット ...
+
+この機能には、変換機能と IC コンセンサスが重要な役割を果たすことがおわかりいただけると思います。変換機能は、レプリカが受信したレスポンスの差異を確実に取り除き、異なるレプリカ上の変換されたレスポンスが全く同じになるようにすることで、コンセンサスが呼び出し側の Canister に合意されたレスポンスを提供することを可能にします。レプリカが受信したレスポンスをコンセンサスで実行することで、どのレプリカもスマートコントラクト Wasm の実行環境に対して同じレスポンスを提供することを保証しています。
```
[specification](https://dfinityjp.netlify.app/docs/current/references/ic-interface-spec/#ic-http_request)の方にも言及ありますが、transform function は、（レスポンスを変換するための）関数です。
```suggestion
この機能には、変換関数と IC コンセンサスが重要な役割を果たすことがおわかりいただけると思います。変換関数は、レプリカが受信したレスポンスの差異を確実に取り除き、異なるレプリカ上の変換されたレスポンスが全く同じになるようにすることで、コンセンサスが呼び出し側の Canister に合意されたレスポンスを提供することを可能にします。レプリカが受信したレスポンスをコンセンサスで実行することで、どのレプリカもスマートコントラクト Wasm の実行環境に対して同じレスポンスを提供することを保証しています。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+
+上の図は、Canister がどのように機能と相互作用し、サブネットレプリカが外部サーバーと通信するパターンを概観しています。
+
+要約すると、HTTP リクエストを実行するために、各レプリカは、外部の Web サーバーから受け取った HTTP レスポンスの（オプションで変換された）インスタンスを Internet Computer のコンセンサス層を介してプッシュします。サブネットのレプリカは、レプリカが受け取ったすべてのサーバーレスポンスに基づいて、Canister に提供するレスポンスに合意できるようにします。オプションの変換により、サーバーから各レプリカで受信したレスポンスが部分的に異なる場合、それらの違いは排除され、同じ変換されたレスポンスがすべての（正直な）レプリカのコンセンサスに提供されることが保証されます。これにより、すべてのレプリカにおいて、全く同じレスポンス（または何も違わない）が Canister の実行に使用されることが保証され、それによってこの機能を使用するときに不一致が起こらず、サブネット ...
```
- オプションの変換、の意味が理解しづらいと感じましたので意訳しました。ご意見あればお願いします
- (or none at all) is used は、「または何も使用されない」と訳すのが良いと思います。つまり、レスポンスが異なっていれば、コンセンサスには何も使われないことを保証していると言っています。（レスポンスがバラバラなのにコンセンサスに使われてしまっては困るので）
```suggestion
要約すると、HTTP リクエストを実行するために、各レプリカは、外部の Web サーバーから受け取った HTTP レスポンスの（変換関数を指定した場合には変換された）インスタンスを Internet Computer のコンセンサス層を介してプッシュします。サブネットのレプリカは、レプリカが受け取ったすべてのサーバーレスポンスに基づいて、Canister に提供するレスポンスに合意できるようにします。変換関数の指定により、サーバーから各レプリカで受信したレスポンスが部分的に異なる場合、それらの違いは排除され、同じ変換されたレスポンスがすべての（正直な）レプリカのコンセンサスに提供されることが保証されます。これにより、すべてのレプリカにおいて全く同じレスポンスが Canister の実行に使用される（または何も使用されない）ことが保証され、それによってこの機能を使用するときに不一致が起こらず、サブネットの複製ステートマシンの性質が維持されることが保証されます。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+### トラストモデルとプログラミングモデル
+
+Canister HTTP アウトコールのトラストモデルは，呼び出された HTTP サーバーと IC のトラストモデルに基づいています：
+* HTTP サービスは正直であると仮定します。そうでなければ，サブネットのどの呼び出し側レプリカに対しても，任意のレスポンスを提供することができません。これは、Web 2.0 サービスやオラクルからサービスを呼び出す場合と同様で、ブロックチェーンからサービスを呼び出す場合もこれと同じシュチュエーションです。
```
otherwise ... は「さもなければ〇〇である」で、必ずしも否定の意味の節がくるとは限りません。文意が読み取りづらいと思いますので主語を括弧書きで追記しました。
```suggestion
* HTTP サービスは正直であると仮定します。そうでなければ、（HTTP サービスは）サブネットのどの呼び出し側レプリカに対しても、任意のレスポンスをできることになります。これは、Web 2.0 サービスやオラクルからサービスを呼び出す場合と同様で、ブロックチェーンからサービスを呼び出す場合もこれと同じシュチュエーションです。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+
+Canister HTTP アウトコールのトラストモデルは，呼び出された HTTP サーバーと IC のトラストモデルに基づいています：
+* HTTP サービスは正直であると仮定します。そうでなければ，サブネットのどの呼び出し側レプリカに対しても，任意のレスポンスを提供することができません。これは、Web 2.0 サービスやオラクルからサービスを呼び出す場合と同様で、ブロックチェーンからサービスを呼び出す場合もこれと同じシュチュエーションです。
+* IC のトラストモデルは、少なくとも $2/3$ のレプリカが正直であると仮定しています。この場合、正直なレプリカは（オプションの変換後に）同じレスポンスを取得し、そのレスポンスを呼び出し元の Canister に提供し返せるように、コンセンサスを得ることができます。
```
```suggestion
* IC のトラストモデルは、少なくとも $2/3$ のレプリカが正直であると仮定しています。この場合、正直なレプリカは（オプショナルである変換後に）同じレスポンスを取得し、そのレスポンスを呼び出し元の Canister に提供し返せるように、コンセンサスを得ることができます。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+Canister HTTP アウトコールのトラストモデルは，呼び出された HTTP サーバーと IC のトラストモデルに基づいています：
+* HTTP サービスは正直であると仮定します。そうでなければ，サブネットのどの呼び出し側レプリカに対しても，任意のレスポンスを提供することができません。これは、Web 2.0 サービスやオラクルからサービスを呼び出す場合と同様で、ブロックチェーンからサービスを呼び出す場合もこれと同じシュチュエーションです。
+* IC のトラストモデルは、少なくとも $2/3$ のレプリカが正直であると仮定しています。この場合、正直なレプリカは（オプションの変換後に）同じレスポンスを取得し、そのレスポンスを呼び出し元の Canister に提供し返せるように、コンセンサスを得ることができます。
+不誠実なサーバー（レプリカ）は、簡単にリクエストを失敗させたり、間違ったデータを提供したりすることになります。また、$1/3$ 以上の不誠実なサブネットレプリカでリクエスト自体を失敗させることができます。2/3$以上のレプリカが危険にさらされると、間違ったデータを提供してしまうことになります。
```
ここでのcanは、「不誠実なサーバーがリクエストを失敗させることができる」という意味で使われています
```suggestion
不誠実なサーバー（レプリカ）は、簡単にリクエストを失敗させたり、間違ったデータを提供することができます。また、$1/3$ 以上の不誠実なサブネットのレプリカはリクエスト自体を失敗させることができます。間違ったデータを提供するためには、$2/3$ 以上のレプリカの共謀が必要です。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+* IC のトラストモデルは、少なくとも $2/3$ のレプリカが正直であると仮定しています。この場合、正直なレプリカは（オプションの変換後に）同じレスポンスを取得し、そのレスポンスを呼び出し元の Canister に提供し返せるように、コンセンサスを得ることができます。
+不誠実なサーバー（レプリカ）は、簡単にリクエストを失敗させたり、間違ったデータを提供したりすることになります。また、$1/3$ 以上の不誠実なサブネットレプリカでリクエスト自体を失敗させることができます。2/3$以上のレプリカが危険にさらされると、間違ったデータを提供してしまうことになります。
+
+この機能のプログラミングモデルは、HTTP アウトコールを行う Canister が *HTTP クライアント*として動作し、様々なヘッダーを解釈し、それに応じて動作する、よくある一般的なケースようなものと似ています。API リクエストのような使用例では、これはかなり簡単で、標準的なケースでは多くの特別な考慮は必要ありません。IC プロトコルスタックは、概念的には、Canister と従来の HTTP サーバーの間の通信パイプとみなすことができ、HTTP レスポンスが合意を経て、すべての正直なレプリカが実行時にまったく同じレスポンスを受け取るようにする、つまり、呼び出し側の Canister が合意されたレスポンスを受け取るようにするものです。
```
needs to の文意を反映しました。後半の文は長いので切って良いと思います。
```suggestion
この機能のプログラミングモデルでは、HTTP アウトコールを行う Canister が *HTTP クライアント*として動作します。つまり、一般的なケースにおいては、様々なヘッダーを解釈し、それに応じて動作する必要があります。API リクエストのような使用例では、これはかなり簡単で、標準的なケースでは多くの特別な考慮は必要ありません。IC プロトコルスタックは、概念的には、Canister と従来の HTTP サーバーの間の通信パイプとみなすことができ、HTTP レスポンスが合意を経て、すべての正直なレプリカが実行時にまったく同じレスポンスを受け取るようにします。つまり、呼び出し側の Canister が合意されたレスポンスを受け取るようにするものです。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+
+基本的なアーキテクチャは、チェーン上にオラクルスマートコントラクトが配置されていることです。ユーザーはこのオラクルスマートコントラクトにリクエストを行い、オラクルスマートコントラクトはそのリクエストを一時的に保存します。オラクル、つまり Web 2.0 サーバーは、クエリやアップデートコールを使ってオラクルスマートコントラクトと通常のやり取りをすることで、保存されたリクエストを取得します。次に、オラクルは通常の HTTP コールによって Web 2.0 の世界でリクエストを実行し、ブロックチェーンとの通常の対話パラダイム（アップデートコール）を使用して、オラクルスマートコントラクトにレスポンスを返します。その後、オラクルスマートコントラクトは、標準的なオンチェーン相互作用を使用して、呼び出したスマートコントラクトにレスポンスを提供し返します。
+
+オラクルアーキテクチャでは、単一のオラクルがオリジナルのリクエストを実行するかもしれないし、オラクルの分散ネットワークがそれぞれリクエストを発行し、合意したレスポンス（証拠を含む）をオラクルコントラクト、最終的には呼び出し側の Canister に提供するかもしれません。このように、オラクルの世界でアウトコールと同じ結果を効果的に達成することができますが、そのサービスの対価を求めるひとつ以上の外部パーティを必要とし、より複雑なアーキテクチャになります。したがって、HTTP アウトコールは多くの関連するユースケースでオラクルに取って代わり、より強力なトラストモデルで、より低い手数料とより低いリクエスト実行レイテンシーでそうすることができると仮定することができます。
```
少し言い換えました
```suggestion
オラクルアーキテクチャでは、単一のオラクルがオリジナルのリクエストを実行するかもしれないし、オラクルの分散ネットワークがそれぞれリクエストを発行し、合意したレスポンス（証拠を含む）をオラクルコントラクト、最終的には呼び出し側の Canister に提供するかもしれません。このように、オラクルの世界でアウトコールと同じ結果を効果的に達成することができますが、そのサービスの対価を求めるひとつ以上の外部パーティを必要とし、より複雑なアーキテクチャになります。したがって、HTTP アウトコールは多くの関連するユースケースでオラクルに取って代わり、より強力なトラストモデルで、より低い手数料とより低いリクエスト実行レイテンシーで同様のことができると考えることができます。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+
+### 開発者の便益
+
+開発者にとって、Canister HTTP リクエスト機能は、 IC 以外にトラストする必要がなく、どの当事者（オラクル）を信頼するかを決定する必要がないという利点があります。トラストの前提を減らしより良い分散化を得るために、単一のオラクルを使用するか、複数のオラクルを使用するかを決定する必要がなく、HTTP アウトコールでこれらを得ることができるのです。HTTP アウトコールのコストは、確立されたオラクル・プロバイダーにそのサービスと関連するイングレス・コストを支払うよりもはるかに低いと思われます。したがって、HTTP コールを使用することには明らかな経済的利点があります。分かりにくい信頼性の判断から解放された開発者は、ビジネスロジックに集中し、必要な HTTP コールを行う（そして対応する変換関数を書く）だけでよく、より自然な形で開発することができます。
```
IC that they already need to trust anyway.　の文意を追加しました
```suggestion
開発者にとって、Canister HTTP リクエスト機能は、いずれにせよ信頼が必要となる IC に加えて、どの当事者（オラクル）を信頼するかを決定する必要がないという利点があります。トラストの前提を減らしより良い分散化を得るために、単一のオラクルを使用するか、複数のオラクルを使用するかを決定する必要がなく、HTTP アウトコールでこれらを得ることができるのです。HTTP アウトコールのコストは、確立されたオラクル・プロバイダーにそのサービスと関連するイングレス・コストを支払うよりもはるかに低いと思われます。したがって、HTTP コールを使用することには明らかな経済的利点があります。分かりにくい信頼性の判断から解放された開発者は、ビジネスロジックに集中し、必要な HTTP コールを行う（そして対応する変換関数を書く）だけでよく、より自然な形で開発することができます。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+
+エンドユーザーは HTTP アウトコールから、様々な方法で利益を得ることができます。
+* *より強力なトラストモデル：* エンドユーザーは、さらなる障害点になるかもしれない付加的なパーティーに依存しないことで、より強力なセキュリティを得ることができ、それによってより良い分散化の恩恵を受けることができます。オラクルサービスのようなサードパーティが関与しないので、サブネットのレプリカと呼び出される外部サービスだけを使った我々の HTTP アウトコールは分散化の最たるものです。
+* *低料金：* オラクルのミドルパーティがカットされるため、ユーザーは通常より安いサービスを得ることができます。HTTP を直接呼び出す場合のレイテンシーは、オラクルコントラクトに（Xnet）リクエストを出し、ポーリングののちにサービスされる外部のオラクルサービスよりも低く、Canister と対話する際のユーザーエクスペリエンスが向上します。オラクルプロバイダが関連するすべてのサブネットにオラクルスマートコントラクトをセットアップしないといけなくなる（複数の） Xnet リクエストの場合に顕著で、これは IC のサブネットアーキテクチャの原則とやや矛盾することにもなります。
```
まず、

> This is more pronounced for Xnet requests

This はレイテンシを指しているので、「（複数の）Xnet リクエストがある場合レイテンシーがさらに顕著になる」というのがメインの主語述語です。

> Xnet requests, which are required unless the oracle provider sets up an oracle smart contract on every relevant subnet

こちらの文では、どのような場合にXnet リクエストが必要なのか述べています。
「オラクルプロバイダが関連するすべてのサブネットにオラクルスマートコントラクトをセットアップしない限り（複数の）Xnet リクエストが必要」と言っています。
言い換えると、どこかのサブネットに代表してオラクルスマートコントラクトをセットアップした場合にはXnet リクエストが必要になり、全てのサブネットに対してオラクルスマートコントラクトをセットアップした場合にはXnet リクエストは必要ないということです。

> which would somewhat contradict the subnet architecture principle of the IC.

which は前の "オラクルプロバイダが関連するすべてのサブネットにオラクルスマートコントラクトをセットアップすること" を指しています。
つまり、「関連するすべてのサブネットにオラクルスマートコントラクトをセットアップすることがIC のサブネットアーキテクチャ原則に反している」と言っています。

以上をまとめて以下のように訳しました。
> IC のサブネットアーキテクチャの原則といくらか矛盾する、オラクルプロバイダが関連するすべてのサブネットにオラクルスマートコントラクトをセットアップするということをしない限り、（複数の）Xnet リクエストが必要になり、レイテンシーがさらに顕著になります。

ICのサブネットアーキテクチャでは基本的にサブネットを意識せずに使える（そのため"the" blockchainである）のに対し、「オラクルプロバイダが関連するすべてのサブネットにオラクルスマートコントラクトをセットアップしないといけなくなる」というのは、サブネットを強く意識した構成にしなければいけなくなるので、サブネットへの意識が必要ないという原則に対して矛盾が生じるということなのではないかと受け取りました（例えば、サブネットが追加された時に追加されたサブネットに対して何かしら外部サービスでセットアップが必要な場合、アーキテクチャ原則に反する）

```suggestion
* *低料金：* オラクルのミドルパーティがカットされるため、ユーザーは通常より安いサービスを得ることができます。
* *低レイテンシー：* HTTP を直接呼び出す場合のレイテンシーは、オラクルコントラクトに（Xnet）リクエストを出し、外部のオラクルサービスによってポーリングののちにサービスされる場合に比べて低く、Canister と対話する際のユーザーエクスペリエンスが向上します。IC のサブネットアーキテクチャの原則といくらか矛盾する、オラクルプロバイダが関連するすべてのサブネットにオラクルスマートコントラクトをセットアップするということをしない限り、（複数の）Xnet リクエストが必要になり、レイテンシーがさらに顕著になります。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+
+サブネットのレプリカが受け取るレスポンスは、各レスポンスが同じ変換関数を受けることができ、変換の結果がどのレプリカでも等しいという意味で、&ldquo;同様&rdquo;でなければなりません。したがって、この文脈で同様とは、我々が関心を持ついくつかの核となる情報がすべてのレスポンスで等しく、他の部分は異なっていても、レスポンスには関係ないことを意味します。HTTP API の世界でよくある設定は、レスポンスは構造的に等しいが、レスポンスに含まれるタイムスタンプや識別子などのために、あるフィールドが異なるというものです。
+
+### POST リクエストはべき乗でなければならない
```
typoeです
```suggestion
### POST リクエストはべき等でなければならない
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+
+### POST リクエストはべき乗でなければならない
+
+この機能を実装する方法は、すべての HTTP リクエストをサブネットのすべてのレプリカが行う必要があることを意味します。これは `GET` だけでなく `POST` リクエストにも当てはまり、後者では `GET` リクエストにはない課題が生じ、基本的な HTTP の原則に従って適切に実装されるならば、その課題はべき等です。べき等とは、リクエストが複数回行われた場合、同じ結果をもたらし、サーバーの状態を変化させないということです。べき等は `POST` リクエストには適用されません。つまり、さらなる予防措置がなければ、Canister から行われた `POST` は、サブネットのレプリカの数を $n$ として、$n$ 回、呼ばれたサーバーの更新につながるリクエストになる可能性があるということです。この動作は明らかに意図したものではありませんし、ほとんどのシナリオで受け入れられるものでもあり ...
```
読みやすさのため括弧書きで補足を入れつつ言い換えました。
また、文が長い箇所は切った方が読みやすいと思いました
```suggestion
この（HTTP アウトコールの）機能の実装方法が意味するところは、すべての HTTP リクエストをサブネットのすべてのレプリカが行う必要があるということです。これは `GET` だけでなく `POST` リクエストにも当てはまり、後者では `GET` リクエストにはない課題が生じます。基本的な HTTP の原則に従って適切に実装されるならば、その課題はべき等です。べき等とは、リクエストが複数回行われた場合、同じ結果をもたらし、サーバーの状態を変化させないということです。べき等は `POST` リクエストには適用されません。つまり、さらなる予防措置がなければ、Canister から行われた `POST` は、サブネットのレプリカの数を $n$ として、$n$ 回、呼ばれたサーバーの更新につながるリクエストになる可能性があるということです。この動作は明らかに意図したものではありませんし、ほとんどのシナリオで受け入れられるものでもありません。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+
+### 不正なレプリカ
+
+すでに述べたように、不正なレプリカはより深刻な問題です。そのようなレプリカは、正直なレプリカによって実装された Canister が気づいていない任意の `POST` リクエストを、いつでもサービスに対して行うことができます。つまり、外部サーバーを認証するために API キーやパスワードを保存している場合、不正なレプリカが保存されている平文の認証情報を使ってサーバーを認証してしまうという問題に直面するのです。
```
generalなので一般的とするのが良いと思いました
```suggestion
すでに述べたように、不正なレプリカはより一般的な問題です。そのようなレプリカは、正直なレプリカによって実装された Canister が気づいていない任意の `POST` リクエストを、いつでもサービスに対して行うことができます。つまり、外部サーバーを認証するために API キーやパスワードを保存している場合、不正なレプリカが保存されている平文の認証情報を使ってサーバーを認証してしまうという問題に直面するのです。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+
+リクエストが適切に実行されたかどうかをチェックするために、`GET` を介して安全な *読み直し* 操作を行うことは、課題のいくつかを軽減するのに役立ちます。しかし、不正なレプリカが状態の全く異なる部分を更新することは、この方法では信頼性をもって検出できません。
+
+### 不正なレプリカ
```
特に違和感ありません

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+### Canister HTTP コールのコーディングのレシピ
+
+次の &ldquo;レシピ&rdquo; は、新しい Canister HTTP アウトコールタイプの実装に取り組むための最良の方法のブループリントになります。
+* &ldquo;手動&rdquo; 例えば、`curl` ツールを使って、サブネットのレプリカによって行われることをエミュレートするために、短い時間枠（1-2秒）内に関心のある同じ HTTP リクエストを2回、ちょうどより小さい $n$ で行います。
```
文の一部なので「手動で、」とするのが良いと思います。curlコマンドで手打ちすることを言っています。
justはここではちょうどというより、「単純に」くらいの意味です
```suggestion
* &ldquo;手動&rdquo; で、例えば、`curl` ツールを使って、サブネットのレプリカによって行われることをエミュレートするために、短い時間枠（1-2秒）内に関心のある同じ HTTP リクエストを2回、単により小さい $n$ で行います。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+* *SysFatal - Timeout expired:* リクエストは、タイムアウト期間内に実行されない場合、タイムアウトします。これが起こる重要な例のひとつは、コンセンサスを得るために十分な数の等しいレスポンスがないときです。これは、例えば、変換関数が応答のすべての可変部分を考慮に入れて正確に書かれていないときに起きます。
+* *SysTransient - Failed to connect: error trying to connect: tcp connect error: Connection refused (os error 111):* このエラーは、相手サーバーとの TCP 接続が確立できなかったことを示します。
+* *CanisterReject - http_request request sent with 0 cycles, but ... cycles are required:* 少なくとも必要な量の Cycle が、サブネットで満たされるために要求と共に送信される必要があります。
+* *CanisterReject - max_response_bytes expected to be in the range [0..2097152], got …:* このエラーは、サーバーから受信したネットワークレスポンスが大きすぎることを示します。このエラーは、サーバーから受け取ったネットワーク応答が大きすぎることを示します。これは、応答サイズが過小評価され、`max_response_bytes` の値が低く設定された場合に発生します。
```
重複しています
```suggestion
* *CanisterReject - max_response_bytes expected to be in the range [0..2097152], got …:* このエラーは、サーバーから受信したネットワークレスポンスが大きすぎることを示します。これは、応答サイズが過小評価され、`max_response_bytes` の値が低く設定された場合に発生します。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+* *SysFatal - Transformed http response exceeds limit: 2045952:* このエラーは、変換されたレスポンスサイズの制限に達したことを示します。これは現在、HTTPS アウトコール機能のハードレスポンスサイズ制限です。レスポンスサイズは、レスポンスボディとヘッダーに基づいて計算されることに注意してください。
+
+この機能を初めて使う開発者は、最初のうちはある種の問題にぶつかる可能性が高く、それは以下のエラーのいずれかとして具体化されます。この機能を使い始める際に、最も顕著であると思われる問題を列挙します。
+* ある特定のタイプの Canister HTTP リクエストがローカルの dfx 環境では動作しても、IC では通常のアプリケーションのサブネットで $n=13$ 個のレプリカが動作するのに対し、ローカル環境では $1$ 個のレプリカが動作するため、動作しない可能性があります。このような問題は、このような呼び出しを開発する際、特に開発者がその機能を扱うのに必要な経験をまだ積んでいない場合に予想されます。ここで予想される主な問題は、変換関数の不足または問題です。なお、dfx 環境と IC へのデプロイの違いは、dfx 環境の動作に起因するものであり、すぐに変わるものではありません。dfx 環境は、ローカルにひとつのレプリカを実行し、エンジニアリングプロセスにおけるすべての長所と短所を伴います。
```
若干読みにくかったので言い換えました
```suggestion
* IC では通常のアプリケーションのサブネットで $n=13$ 個のレプリカが動作するのに対し、ローカル環境では $1$ 個のレプリカが動作するため、ある特定のタイプの Canister HTTP リクエストがローカルの dfx 環境では動作しても、IC では動作しない可能性があります。このような問題は、このような呼び出しを開発する際、特に開発者がその機能を扱うのに必要な経験をまだ積んでいない場合に予想されます。ここで予想される主な問題は、変換関数の不足または問題です。なお、dfx 環境と IC へのデプロイの違いは、dfx 環境の動作に起因するものであり、すぐに変わるものではありません。dfx 環境は、ローカルにひとつのレプリカを実行し、エンジニアリングプロセスにおけるすべての長所と短所を伴います。
```

hoosan commented on docs/developer-docs/integrations/http_requests/http_requests-how-it-works.md:
```diff
@@ -1,3 +1,167 @@
+
+IC の他の機能と同様に、Canister HTTP アウトコール機能は、使用時に課金されます。現在の価格設定は、HTTP リクエストに対して $400$ M Cycle の基本料金に加え、リクエスト 1バイトと `max_response_bytes` バイトごとに $100$ K Cycle を課金するよう定義されています。リクエストごとの固定コストと、ヘッダーなどによる HTTP リクエストのオーバーヘッドのため、アプリケーションの観点から実行可能であれば、より小さなリクエストを多数回行うのと同じ情報を取得するために、より大きなレスポンスを持つより少ないリクエストを行うことがコストの観点から有利です。呼び出しで提供される Cycle は、リクエストのコストをカバーするのに十分でなければならず、過剰な Cycle は呼び出し元に返されます。
+
+現在の価格設定はかなり保守的（高価）であると定義されており、価格は今後、価格モデルの更新が導入されることで変更される可能性があります。しかし、ファイナンス API への問い合わせに使用されるような小さなレスポンスを伴う HTTP アウトコールは、USD セントの端数しかかからず、ほとんどのブロックチェーンでオラクルがコールする際に課される手数料よりも大幅に安いことに注意してください。
```
ここでのnote that は注意すべきというというより、他に比べて優れているので「頭に入れておいて」という意味で使われています。あまり適当な訳が浮かびませんでしたが、言い換えました。
```suggestion
現在の価格設定はかなり保守的（高価）であると定義されており、価格は今後、価格モデルの更新が導入されることで変更される可能性があります。しかし、ファイナンス API への問い合わせに使用されるような小さなレスポンスを伴う HTTP アウトコールは、USD セントの端数しかかからず、ほとんどのブロックチェーンでオラクルがコールする際に課される手数料よりも大幅に安いということは述べておきましょう。
```

hoosan reviewed (changes requested):
翻訳ありがとうございます。レビューしましたのでご確認お願いします。
 &ldquo; に関しては、適切に表示できているので問題ないと思いました

hokosugi committed:
Apply suggestions from code review

レビュー、ありがとうございます！
大変参考になりました。特に、サブネットアーキテクチャの話は勉強になりました。
日本語版ドキュメントがとても読みやすくなっていると実感しています。

Co-authored-by: hoosan <40290137+hoosan@users.noreply.github.com>
